docs/design-documents/20221024-stream-inspector.md

```diff
@@ -0,0 +1,90 @@
+# Stream Inspector
```

Comment:
(Adding a comment on this line just so we can have a discussion, otherwise it's hard to follow)

> > but we currently can't address running nodes, they are spawned and run but not stored anywhere.
>
> How about [this](https://github.com/ConduitIO/conduit/blob/main/pkg/pipeline/instance.go#L60) though?

Oh, I forgot we had that 🙂 That pretty much solves the problem of addressing a node, the nodes that represent a certain entity (source, destination, processor) have the same ID as that entity, so it should be easy to fetch the right node. The other side of the problem is how we attach to the stream (and detach afterwards), but that's something we can figure out when we start implementing it 👍

Reply:
We can take inspiration from [github.com/eapache/channels#Tee](https://pkg.go.dev/github.com/eapache/channels#Tee) or [`io.TeeReader`](https://pkg.go.dev/io#TeeReader).

Reply:
Having the ability to tee any output, without interruption, can be pretty nice.

Reply:
> The other side of the problem is how we attach to the stream (and detach afterwards), but that's something we can figure out when we start implementing it +1

When you say "stream" what exactly do you mean?

Reply:
I meant how we listen to messages flowing through a channel between two nodes and how this affects the pipeline performance.

That opens another question - what happens if the pipeline inspector is slower than the rate of messages that flow through the pipeline? Does it cause the pipeline to slow down (i.e. it applies backpressure)? Or should messages just be dropped if the inspector can't handle them in time? It sounds to me like the inspector should guarantee that it will return all messages that flow by in the correct order, but that necessarily means the pipeline performance can be impacted if the inspector is too slow.

Reply:
Interesting question! :+1: 

I think we need to explicitly mention that the inspector is expected to have an insignificant impact on the pipeline, if at all. There's already a requirement that the inspector has insignificant resource usage. In other words, we make our best efforts to make inspector lightweight. If an inspection is slow, it's the client's fault.:) With that, what we could is to keep a copy of the records, only for the inspector. 

> what happens if the pipeline inspector is slower than the rate of messages that flow through the pipeline?

If the pipeline inspector or the client is slow for whatever reason, and we start dropping messages, that might make someone think that messages are being lost for real. So we should either send a warning to the client that messages are going to be dropped, or we close the inspection.

Reply:
> With that, what we could is to keep a copy of the records, only for the inspector.

Having a buffer might help, although it doesn't prevent the inspector from impacting the performance of a pipeline (the buffer can't be unbounded).

> If the pipeline inspector or the client is slow for whatever reason, and we start dropping messages, that might make someone think that messages are being lost for real.

Good point. The more I think about dropping messages the more it's clear we should not do that. I (as a user) would rather have a slower pipeline and know I see everything that flows by than get an incomplete stream of messages.

> we make our best efforts to make inspector lightweight. If an inspection is slow, it's the client's fault.

That sounds good to me.

> we need to explicitly mention that the inspector is expected to have an insignificant impact on the pipeline

That doesn't sound good to me 🙃 I mean we can say that _generally_ this is the case, but we need to be transparent that it can impact the pipeline performance if the pipeline inspector client doesn't fetch messages fast enough, in that case we will apply backpressure.

Reply:
> Having a buffer might help, although it doesn't prevent the inspector from impacting the performance of a pipeline **(the buffer can't be unbounded)**.

Why not though?

> that it can impact the pipeline performance if the pipeline inspector client doesn't fetch messages fast enough, in that case we will apply backpressure.

IMO, troubleshooting a problem (i.e. inspecting a pipeline) shouldn't introduce another problem (pipeline getting slower). The pipeline and the inspector should be as isolated one from another as possible, so that we're as confident as possible that bugs in one won't cause bugs in the other. For me personally, it would give much more freedom to actually inspect and not be afraid I'll make things worse if I need to look around or do some tests.

Reply:
> Physics? smile We can potentially fill up the memory, possibly causing the system to use swap space. That will cause a performance impact far greater than just the one pipeline. I'd rather try to restrict the possible impact.

I'm looking back at your original message and am surprised to see "unbounded", I was thinking: why prevents us from limiting the buffer?:)

> after the buffer is full it will start to throttle the speed of the pipeline though.

Personally, I wouldn't like to see that as someone inspecting a pipeline. A hanging UI/CLI, an incautious developer, a network latency etc. could cause the inspection to be slow, but I'd never expect something like that to throttle a pipeline. The more connected the pipeline and an inspector are, the higher are the changes for bugs and performance degradations.

Reply:
I want to circle back at the inspector being non-blocking only, I think inserting an inspector which allows for:
1. see data
2. data ok/not-ok -> ack/nack from the inspector

Could be useful in some interesting scenarios where but I think it will be less useful if we cannot put the data back.

Reply:
@lyuboxa that's what I was afraid of 😆 We need to define the purpose of the inspector - I see it as a tool to inspect and debug records, not more, not less. Not for filtering or transforming records - that's a processor. It's also not a tool for consuming records and writing them to some resource - that's a destination. People can abuse the tool, but we should not design for that.

What you described would be a pretty interesting processor though, one where a human would be involved to confirm or reject a record. Could be used to do things like fraud detection 🙂

Reply:
> What you described would be a pretty interesting processor though, one where a human would be involved to confirm or reject a record. Could be used to do things like fraud detection slightly_smiling_face

Hmm you might be onto something, high latency human processor ;-)

Reply:
@lyuboxa What would be those scenarios (manually acking/nacking messages and inserting data into a pipeline)?

Reply:
Here's the conclusion from today's team sync: https://github.com/ConduitIO/conduit/pull/692/files#diff-88c2b66f0b7457fc1c81362cacfa6ca628eb888f2326028c0abd9d5f538a79d4R123-R131